### PR TITLE
Added DataTreeConverter tests

### PR DESCRIPTION
**Pull request by Earthmark**
Added initial tests for the data tree, these avoid file access.

Also added IEnumerable to DataTreeDictionary, this allows collection initializers on the object, making test cases easier to generate.

Before
```csharp
var element = new DataTreeDictionary();

element.Add("Id", i);
element.Add("Name", "element-" + i);
element.Add("Position", new float3(i * 0.5f, i * 0.25f, i * 0.125f));
element.Add("Enabled", (i & 1) == 0);
element.Add("Weight", i * 1.5);

```

After
```csharp
var element = new DataTreeDictionary
{
    { "Id", i },
    { "Name", "element-" + i },
    { "Position", new float3(i * 0.5f, i * 0.25f, i * 0.125f) },
    { "Enabled", (i & 1) == 0 },
    { "Weight", i * 1.5 }
}
```

Enumeration is for the keys on that level of the dictionary, EnumerateTree is for recursion.

### CODE REVIEW

**Comment by Earthmark** [comment]
~~As with https://github.com/Yellow-Dog-Man/Elements.Core/pull/28 , once I hear back about what the test format is supposed to be, I'll update the test names to match the expected style. For now they're in the style of the local repo.~~

This was lightly discussed in https://github.com/Yellow-Dog-Man/Elements.Style/issues/11 , the takeaway was there isn't really a concrete existing style. The pr was adjusted to follow the standard .net test style.

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        return new DataTreeDictionary
+        {
+            { "Items", tree }
+        };
+    }
+
+    [TestMethod]
+    public void EnumerateKnownChildSequence()
```
Comment (Earthmark) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs]: This was the test added to verify https://github.com/Yellow-Dog-Man/Elements.Core/pull/33 , this is possibly not a useful test though. Some of the collections being enumerated are dictionaries.

If the order isn't important this can be changed to just ensure all items are in the enumeration, instead of caring about the order.

**Comment by Earthmark** [comment]
More tests around historic file parsing should be added in the future, a lot of the DataTreeConverter is repeated sections of file saving code, but it's difficult to simplify without historic edge cases.

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+[TestClass]
+public class DataTreeConverterTests
+{
+    [DataRow("asset.lz4bson", true)]
+    [DataRow("asset.7zbson", true)]
+    [DataRow("asset.brson", true)]
+    [DataRow("ASSET.LZ4BSON", true)]
+    [DataRow("asset.bson", false)]
+    [DataRow("asset.json", false)]
+    [DataRow("asset.frdt", false)] // TODO: This should be true, and is probably a bug.
```
Comment (Frooxius) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs]: We generally don't use uncompressed versions, so I think it's ok.

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    public record TreeCase(string Name, DataTreeDictionary Tree)
+    {
+        public override string ToString() => Name;
+    }
+
+    public static IEnumerable<TreeCase> HappyLittleTrees
+    {
+        get
+        {
+            // Commented types do not come through as the same type, ints are promoted to longs, floats to doubles.
```
Comment (Frooxius) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs]: Hmm... is this something that can also be replicate with Resonite itself? Say you create ValueField<ulong> and put MaxValue in it, save it and then it doesn't load correctly?

Comment (Earthmark) [reply Elements.Core.Tests/Encoding/DataTreeConverterTests.cs]: In resonite it appears to work correctly, 

Image

In my own testing though, if I turn that off the value that comes back as -1 (int64). I'm assuming some layer converting from the data tree to the component type does a reinterpert cast?

So at the data tree layer this looks suspect as heck, but higher up the type gets converted back so it's fine.

Image

Comment (Earthmark) [reply Elements.Core.Tests/Encoding/DataTreeConverterTests.cs]: As per the discussion in https://github.com/Yellow-Dog-Man/Elements.Core/issues/39, I'll see about swapping this to Verify to see if this is a bit easier to understand. That may express the long-double conversion more easily.

Comment (Earthmark) [reply Elements.Core.Tests/Encoding/DataTreeConverterTests.cs]: Turns out this gets messy, Verify doesn't include type names on the value fields, even if `settings.AddExtraSettings(s => s.TypeNameHandling = TypeNameHandling.All|Auto)` is used. It only includes type names on the object types.

For now I'll leave this as is then, I personally feel it's important to show the type-promotion, but I can shift it in either direction if poked.

Another test should be added at some point anyways to verify the Add methods all work.

**Comment by Earthmark** [comment]
This got a rather massive refresh, effectively consider it a new review.

I actually pointed claude at being a reviewer and it was quite unhappy with my lack of case covering; especially when told to follow the standard.

This now adds a few other test files as well, but I can split them off if requested.

File: Elements.Core/Encoding/DataTreeSerializer/DataTreeConverter.cs [hunk 4]
```diff
@@ -38,17 +38,17 @@ static DataTreeConverter()
...
         public enum Compression
         {
-            None,
-            LZ4,
-            LZMA,
-            Brotli,
+            None = 0,
+            LZ4 = 1,
+            LZMA = 2,
+            Brotli = 3,
```
Comment (Earthmark) [inline Elements.Core/Encoding/DataTreeSerializer/DataTreeConverter.cs:44]: I noticed these were not tagged, but if this order changed or was added in the middle, this would possibly break all manifests. So it's a bit stronger now!

This should probably get a comment in the source.

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+            { "NegZero", double.NegativeZero }
+        });
+        var leaf = (DataTreeValue)dict["NegZero"];
+
+        Assert.IsTrue(double.IsNegative(leaf.Extract<double>()));
+    }
+
+    #region Serialization Unstable Trees
+
+    public static IEnumerable<TreeCase> HappyLittleSerializationUnstableTrees
```
Comment (Earthmark) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs:151]: I was getting frustrated at seeing the unused test sections due to their serialization stability, so the fixture below was added. And in that a few odd things fell out like the ulong extract thing.

Additional tests should be added to verify more vector types, and the rest of the generated types probably.

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+            { "Bool", true },
+            { "Int", 42 },
+            { "Long", -9000000000L },
+            { "Float", 1.5f },
+            { "Double", -2.25 },
+            { "String", "sample" },
+            { "NullString", null as string },
+            { "Url", new Uri("https://example.com/asset") },
+            // NOTE: This stores the enum as a string, a hard-typed extractor is only able to unpack this correctly.
+            { "Enum", DataTreeConverter.Compression.LZMA },
```
Comment (Earthmark) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs:178]: This one was extra strange, the add call formats this to a string, but you can create a DataTreeValue holding an enum directly that's invalid. 

I didn't test the invalid case, I hope it doesn't come up in prod

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+        return new TreeCase("Deeply Nested Tree", root);
+    }
+
+    /// <summary>
+    /// This tree is a landmine, be very careful using it.
+    /// Assert.AreEquivalent does not support this depth.
+    /// </summary>
+    static TreeCase VeryDeeplyNestedTree()
+    {
+        // Somewhere between 1000 and 2000 causes a stack overflow on save.
```
Comment (Earthmark) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs:336]: The recursion limit should probably be re-enstated, just at around a larger limit. Stack overflows are scary

File: Elements.Core/Encoding/DataTreeSerializer/DataTreeConverter.cs [hunk 8]
```diff
@@ -38,17 +38,17 @@ static DataTreeConverter()
...
         }
 
         public static bool IsSupportedFormat(string file)
         {
             var ext = Path.GetExtension(file).ToLower();
 
-            return ext == ".7zbson" || ext == ".lz4bson" || ext == ".brson" || ext == "frdt";
```
Comment (Earthmark) [inline Elements.Core/Encoding/DataTreeSerializer/DataTreeConverter.cs:51]: This was a bug that could never be hit. The above get extension call always prefixed the string with a ".".

File: Elements.Core/Encoding/DataTreeSerializer/Generic Interface/DataTreeDictionary.cs [hunk 9]
```diff
@@ -1,11 +1,12 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
 
 namespace Elements.Core
 {
-    public partial class DataTreeDictionary : DataTreeNode
+    public partial class DataTreeDictionary : DataTreeNode, IEnumerable<KeyValuePair<string, DataTreeNode>>
```
Comment (Earthmark) [inline Elements.Core/Encoding/DataTreeSerializer/Generic Interface/DataTreeDictionary.cs:9]: This allows collection initializers, which was real useful as the add calls are strongly typed. This made building test cases much cleaner.

File: Elements.Core.Tests/Encoding/Generic Interface/DataTreeValueTests.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    /// The enum is stored as a string, the battery tests extract it as one.
+    /// This hard-typed extractor emulates codegen extractors.
+    /// </summary>
+    [TestMethod]
+    public void ExtractEnum_EnumProvided_ExtractsHardTypedValue()
+    {
+        var expected = DataTreeConverter.Compression.Brotli;
+        var dict = new DataTreeDictionary
+        {
+            { "Enum", expected }
```
Comment (Earthmark) [inline Elements.Core.Tests/Encoding/Generic Interface/DataTreeValueTests.cs:16]: This needed to be added to a dictionary because the dictionary add method contains the logic to convert the enum to the DataTreeValue (which is a string conversion). Extracting it back is on the leaf.

If you build the leaf by the enum, then extract it the extract throws.

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 11]
```diff
@@ -0,0 +1,526 @@
...
+    }
+
+    static readonly ConcurrentDictionary<Type, Func<object, object>> ExtractCache = new();
+
+    static Func<object, object> MakeExtractionMethod(Type type)
+    {
+        // This is sketchy, but work with me...
+        // The extract method does type conversions that are non-trivial.
+        // We don't care how they're done, just that they're 1-1.
+        var method = typeof(DataTreeValue).GetMethod(nameof(DataTreeValue.Extract), types: [])!
```
Comment (Earthmark) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs:504]: These get cached to make it a bit faster. It's still not great but I couldn't think of another way without this being messy as heck.

There might be a way to do this with dynamics, but I didn't wanna look into loading that for one test.

File: Elements.Core.Tests/Encoding/DataTreeConverterTests.cs [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+        Assert.Throws<NotImplementedException>(() => DataTreeConverter.LoadAuto(stream));
+    }
+
+    /// <summary>
+    /// This only tests trees that are somewhat shallow (due to AreEquivalent limitations),
+    /// and that only include longs. The tree deserializes any int as a long, and it gets detected as a diff.
+    /// </summary>
+    [DynamicData(nameof(HappyLittleSerialStableTrees))]
+    [TestMethod]
+    public void SaveLoad_StableTree_TreeSerializesIdentically(TreeCase treeCase)
```
Comment (Earthmark) [inline Elements.Core.Tests/Encoding/DataTreeConverterTests.cs:97]: With assert unstable, I don't think this test is actually testing something we care about anymore.

This verifies the internal format is stable, but we don't really care about that. 

If this goes away though, some of the accessors on the nodes should probably be locked down.

**Comment by Earthmark** [comment]
I just realized Extract is largely called from LoadSaveHelper... there's at least three layers of public interfaces on this thing :| 

Is there a specific layer that external APIs use, or do they use the whole layer cake? If possible I'd like to pull the tests up to whatever layer is used externally; or at least some of the non-exposed layers can be isolated instead.

EDIT: Coder also wraps this interface in some of its calls, that makes 4 layers.

EDIT 2: I did some digging and this layer is used during slot saving at the very least; these tests do cover the interface at a level prod code uses.

**Comment by ProbablePrime** [comment]
As Froox had left comments we should await a response, I will tag him for a review.

**Comment by Earthmark** [comment]
There was a branch mixup on my local client for a moment there, the order of the enumerable implementation in data tree dictionary possibly moved around the ToString method, but past that there was no actual changes.

I was mutating jj history and a sequence of rebases got a bit grumpy with each other.
